### ISSUE TITLE
Using the `omp_gpu_trans` script is resulting in some compiler errors with some socrates files

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#2314]
There are some files in socrates that are failing to compile only after using the `omp_gpu_trans` script on them.

The issue occurs because some static arrays are moved to be dynamic module level declarations from the subroutine level, however these arrays use a `Kind` parameter which is not visible in the module, resulting in failed compilation:

Subroutine declaration in original:
```
 real(kind=realk), dimension(nd_profile) :: x
```

Moved to module:
```
real(kind=realk), allocatable, dimension(:), private :: x
```

Compiler fails as `realk` is not known at the module level.

Easy fix by hand - add the import.

@sergisiso this is the issue i mentioned on teams.

**Comment by LonelyCat124** [comment]
This also happens for another function where
`integer, parameter :: dp = KIND(1.0D0)` is defined in a function - its not moved to the module level with uses of it.

**Comment by sergisiso** [comment]
Can you try in omp_gpu_trans.py, when calling  "normalise_loops(" set "hoist_local_arrays=False" for now?

Also it would be interesting to attempt "kernels_trans.py" which adds OpenACC kernels pragmas and avoid doing complex loop transformations.

**Comment by LonelyCat124** [comment]
I think the issue went away with `hoise_local_arrays=False`
ACC has other issues (some which appear to be compiler bugs).

I think i'll aim to just get the OpenMP code compiling and running for now, and then we can look in more detail at what we need to do to get performance.

**Comment by LonelyCat124** [comment]
`hoist_local_arrays=False` did stop this. The code compiles now (with my own CMakeLists.txt to build it as the Makefile provided causes problems when rebuilding the code, seemingly even when fully cleaned with rm -r *.a as part of it).

The nvidia and intel compilers disagree on some Fortran syntax, but #2202 is the main remaining problem I have with passing through the code and building the code - I assume #2247 may also happen but I've not hit it recently.

GPU performance has generally been bad, there are definitely some inlinable functions in solve_band_k_eqv_scl which prevent target/loop application on double for loops, which I think is the only  way there is enough work to run on the GPU. Those are currently not inlinable according to Andy as I won't be able to use ModuleInlineTrans with nemo yet - once this is running I will try this and then only adding target/loop parallelism to the double loops in the file (which will hopefully avoid all of the initialisation code which isn't really expensive and I think may be causing a lot of memory to be moved to/from the device due to some loops having dependent iterations).

**Comment by sergisiso** [comment]
Pinging @hiker on this since he recently found the same issue but with symbols on the shape/dimension attribute.
> Subroutine declaration in original:
> real(kind=realk), dimension(nd_profile) :: x
> Moved to module:
>real(kind=realk), allocatable, dimension(:), private :: x
> Compiler fails as realk is not known at the module level.

In his PR I recommended the ScopeNode reference_access to manage all this since he was looking for all these symbols. But in this case we want single symbol declaration granularity to know if we can move it. So maybe a symbol does indeed need a reference_access style method? @hiker whats your view?

**Comment by sergisiso** [comment]
> GPU performance has generally been bad, there are definitely some inlinable functions in solve_band_k_eqv_scl which prevent target/loop application on double for loops, which I think is the only way there is enough work to run on the GPU.

If the caller and callee files are both being processed, you can look these functions by name in the script and allow this Calls be inside OpenMP regions and add a "delcare target" to the callee.

**Comment by hiker** [comment]
I don't know much about the inlining implementation, but I have had the same issue for my driver , and I have fixed it (PR not yet ready). I need to look how to integrate this best, @sergisiso 's ticket #2271 (which is a bit more generic) would certainly help, but I might restrict my work to just what I need - I think Sergi's ticket needs more thinking and design (e.g. I think in some cases we might not want that information, in some cased we do want it, so we need a set of options that give us this flexibility). 

With the code I am using in my driver creation, I get the information which symbols are read in the symbol table (i.e. either in array declarations, or assignments), and then use that in the driver, i.e. the writer dumps the symbol into the file, and the driver reads the symbol. Both obviously are then also importing the symbol, which is what I think is needed here.

**Comment by sergisiso** [comment]
Agreed, I updated my ticket description with some these thoughts.

**Comment by LonelyCat124** [comment]
> > GPU performance has generally been bad, there are definitely some inlinable functions in solve_band_k_eqv_scl which prevent target/loop application on double for loops, which I think is the only way there is enough work to run on the GPU.
> 
> If the caller and callee files are both being processed, you can look these functions by name in the script and allow this Calls be inside OpenMP regions and add a "delcare target" to the callee.

The issue is not so simple, essentially there are loops:
```
do i = 1, N
   do = j, N_2
     ...
   end do
   Call my_func(...)
   do j, N_2
     ...
   end do
end do
```
N and N_2 are ~70 and ~700, such that N * N_2 is enough iterations for GPU computation (as the loops themselves are quite lightweight), however neither is individually.

The loops that (maybe) parallelise well (i.e. don't significantly hurt performance right now) all use `collapse`, but this won't even work if I can apply `omp loop` to the outside loop anyway. Potentially I could split this into multiple loops (and alongside inlining or manually doing stuff, also parallelise the call) but I need to check that in more detail.

**Comment by LonelyCat124** [comment]
@sergisiso I see that issue #1928 talks about loop fission for exactly the case I'm coming across here - what was the reason you weren't able to fuse the loops? Did they have different loop bounds (e.g. bound_1 and bound_2) or were they the same symbol but psyclone can't be sure whether the value has changed?

**Comment by sergisiso** [comment]
We still have this case in some places in NEMO, it can be solved by loop fusion or fission but it must be proved that the transformation is valid, which is not always easy and at the time the dependency analysis was less developed than now.

The reason that it was not prioritized was that:
- Memory transfers turned out to be much more important for the first pass. If the outerloop can be brought to the GPU even if not efficiently is still worth it because it means that the data won't be touched and sent to the CPU. And this is paid twice also in the next GPU loop when the data has to be sent back. However it needs a critical mass of loops on the GPU to see the effects because if the next loop is on the CPU anyway, you don't win anything by bringing a loop like the one you shown to the GPU.
- If the outerloop is to small, it can be skipped (e.g. by the name of the routine and loop bound) and put the omp target in the inner loop. In your case this would launch 70 kernels of 700-way parallel. There is a bit of overhead for the small kernels but if then you can proof that the outer loop is iteration independent (which we have a method) you can stick a nowait clause to the inner loop pragma and most of the kernel launching overhead is hidden.
